encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java

```diff
@@ -0,0 +1,169 @@
+      if (!types.isSameType(keyType, elements.getTypeElement("java.lang.String").asType())) {
+        messager.printMessage(
+            Diagnostic.Kind.ERROR,
+            "Cannot encode Maps with non-String keys.",
+            ((DeclaredType) returnType).asElement());
+      }
+    }
+    if (types.isAssignable(
+        returnType, types.erasure(elements.getTypeElement("java.util.Optional").asType()))) {
+      returnType = ((DeclaredType) returnType).getTypeArguments().get(0);
+      getterExpression = getterExpression + ".orElse(null)";
+    }
+
+    return Optional.of(Getter.create(fieldName.get(), getterExpression, returnType));
+  }
+
+  // ownerType = Bar<Optional<Integer>, Boolean>
+  // ownerElement = Bar<T, U>
+  // genericType = U
+  private TypeMirror resolveTypeArguments(DeclaredType ownerType, TypeMirror genericType) {
+
+    if (genericType instanceof DeclaredType) {
+      DeclaredType dType = (DeclaredType) genericType;
+
+      if (dType.getTypeArguments().isEmpty()) {
+        return genericType;
+      }
+
+      TypeElement genericElement = (TypeElement) dType.asElement();
+
+      List<TypeMirror> result = new ArrayList<>();
+      for (TypeMirror typeArgument : dType.getTypeArguments()) {
+        result.add(resolveTypeArguments(ownerType, typeArgument));
+      }
+
+      return types.getDeclaredType(genericElement, result.toArray(new TypeMirror[0]));
+    } else if (genericType instanceof TypeVariable) {
+      TypeVariable tVar = (TypeVariable) genericType;
+      TypeElement ownerElement = (TypeElement) ownerType.asElement();
+
+      int index = ownerElement.getTypeParameters().indexOf(tVar.asElement());
+      if (index == -1) {
+        messager.printMessage(
+            Diagnostic.Kind.ERROR,
+            String.format(
+                "Could not infer type of %s in %s. Is it a non-static inner class in a generic class?",
+                genericType, ownerType));
+        return genericType;
+      }
+      return resolveTypeArguments(ownerType, ownerType.getTypeArguments().get(index));
+    }
+    return genericType;
+  }
+
+  private static Optional<String> inferName(ExecutableElement element) {
+    Encodable.Field annotation = element.getAnnotation(Encodable.Field.class);
+    if (annotation != null && !annotation.name().isEmpty()) {
+      return Optional.of(annotation.name());
+    }
+
+    String methodName = element.getSimpleName().toString();
+    ExecutableType method = (ExecutableType) element.asType();
+
+    if (methodName.startsWith("is")
+        && methodName.length() != 2
+        && method.getReturnType().getKind() == TypeKind.BOOLEAN) {
+      return Optional.of(Character.toLowerCase(methodName.charAt(2)) + methodName.substring(3));
+    }
+
+    if (methodName.startsWith("get")
+        && methodName.length() != 3
+        && method.getReturnType().getKind() != TypeKind.VOID) {
+      return Optional.of(Character.toLowerCase(methodName.charAt(3)) + methodName.substring(4));
+    }
```

Comment:
This is perhaps something good to note in the comments.

Reply:
Good point, updated `@Encodable`s javadoc  to address it.

Reply:
Thanks!